File: src/main/java/nl/mpcjanssen/simpletask/sort/ThresholdDateComparator.java
package nl.mpcjanssen.simpletask.sort;

import android.support.annotation.NonNull;
import com.google.common.collect.Ordering;
import nl.mpcjanssen.simpletask.task.Task;

public class ThresholdDateComparator extends Ordering<Task> {

    @Override
    public int compare(@NonNull Task a, @NonNull Task b) {
        int result;
        if (a.getThresholdDate() == null && b.getThresholdDate() == null) {
            result = 0;
        } else if (a.getThresholdDate() == null) {
            result = 1;
        } else if (b.getThresholdDate() == null) {
            result = -1;
        } else {
            result = a.getThresholdDate().compareTo(b.getThresholdDate());
        }
        return result;
    }
}

File: src/main/java/nl/mpcjanssen/simpletask/sort/ThresholdDateComparator.kt
package nl.mpcjanssen.simpletask.sort

import nl.mpcjanssen.simpletask.task.TodoItem
import java.util.*

class ThresholdDateComparator(val createAsBackup: Boolean) : Comparator<TodoItem> {

    override fun compare(a: TodoItem?, b: TodoItem?): Int {
        if (a === b) {
            return 0
        } else if (a == null) {
            return -1
        } else if (b == null) {
            return 1
        }
        val result: Int

        var dateA = a.task.thresholdDate
        var dateB = b.task.thresholdDate


        // Use create date as threshold date
        // if configured in the settings.
        if (createAsBackup) {
            dateA = dateA ?: a.task.createDate
            dateB = dateB ?: b.task.createDate
        }
        if (dateA == null && dateB == null) {
            result = 0
        } else if (dateA == null) {
            result = 1
        } else if (dateB == null) {
            result = -1
        } else {
            result = dateA.compareTo(dateB)
        }
        return result
    }
}
